### ISSUE TITLE
Package Manager: Conan or vcpkg?

### ISSUE THREAD

**Issue by thejohnfreeman** [XRPLF/rippled#4125]
I'm moving our dependency management to a package manager. My instinct based on my history is to reach for Conan, but @seelabs rightly pointed out that we should first discuss _which_ package manager to use.

In the C++ community, there are presently two front-runners for package management: Conan and vcpkg. I will start the discussion by comparing them. I prefer not to take a bet on a lesser-used package manager, but if someone can make a good case for one in the comments, please be my guest.

I have more experience with Conan, having compared them two years ago and finding it a much better option. Since then, vcpkg has added two features I consider critically important: binary packages and independent versioning. It's worth taking a fresh look at both.

## Binary packages

This feature lets clients download pre-built packages from a remote cache instead of spending time building them locally from source.

[vcpkg](https://github.com/microsoft/vcpkg/blob/master/docs/users/binarycaching.md) takes a [conservative approach](https://github.com/microsoft/vcpkg/blob/master/docs/users/binarycaching.md#implementation-notes-internal-details-subject-to-change-without-notice) to identifying binary builds, while Conan takes a [flexible approach](https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html) that lets it reuse the same binaries in more circumstances. Conan also maintains a publicly available binary cache for [over 100 common configurations](https://github.com/conan-io/conan-center-index/blob/master/docs/supported_platforms_and_configurations.md).

I don't think the options make much difference for this project, though I still think Conan has the better approach and better support. It's unlikely we'll be able to use any of the configurations cached at Conan Center, so we'll have to maintain our own cache. Ripple is already in the practice of using Artifactory (Conan's preferred cache server), and we have even already configured one for this project. vcpkg prefers NuGet, which I'm not sure anyone on the team has experience with.

## Independent versioning

Up until late last year, vcpkg's version model was based on universes, like [Stackage's snapshots](https://www.stackage.org/#about) for Haskell: a universe is a choice of exact versions for every package such that any subset of packages are guaranteed to play nicely together (i.e. build and test with each other). If you wanted to hold off on updating to a new version of package A because of some breaking change, but it had already been updated in the latest universe by the time unrelated package B released critical fixes you've been waiting for, then you'd be out of luck. Conan, on the other hand, works like pip and uses independent versions for all dependencies, leaving it to you to guarantee that your choices of versions cooperate with each other. A few months ago, however, vcpkg [released](https://devblogs.microsoft.com/cppblog/take-control-of-your-vcpkg-dependencies-with-versioning-support/) a feature to enable independent versioning which doesn't seem too complicated.

I consider the options fairly evenly matched on this feature at this point.

## Integration

Going through vcpkg's [integration tutorial](https://vcpkg.io/en/docs/examples/installing-and-using-packages.html), it seems they expect dependents to either (a) use a CMake Config Module installed by the package or (b) write their own Find Module. Conan, on the other hand, will generate [Find Modules](https://docs.conan.io/en/latest/integrations/build_system/cmake/cmake_find_package_generator.html) or [Config Modules](https://docs.conan.io/en/latest/integrations/build_system/cmake/cmake_find_package_multi_generator.html) based on metadata that the recipe author was expected to write, instead of expecting dependents to repeat the same work, possibly with errors. Conan 2.0 is in the works with [significant changes](https://docs.conan.io/en/latest/reference/conanfile/tools/cmake.html) to this pattern that I am not yet familiar with, but do not expect will be harder to work with.

On this issue, I consider Conan a winner.

I feel that these features are the most important to the work of bringing a package manager to rippled, and so I wanted to seed the discussion with them. I have done only shallow research into vcpkg's offerings in these areas. I welcome all questions and contributions in the comments.

**Comment by seelabs** [comment]
@thejohnfreeman Thanks for writing this up. I've toyed with both Conan and vpgk on rippled code and I find them both pretty straight forward, I don't really have a strong preference for one over the other from my (admittedly limited usage). I'm fine with going with Conan.

It would be good if we could figure out if one is more popular than the other in terms of both packages and number of users. I did a quick survey of some packages and they both had them - so I there weren't any red flags for either of them. Still, it would be nice to get a measure of these things.

I'd also factor in what you prefer and are familiar with. If you've been using conan and are happy with it, that definitely matters.

So bottom line: I think I vote for conan; As far as I can tell they are very similar  and the deciding factor is that you are more familiar with/prefer conan.

**Comment by thejohnfreeman** [comment]
Oh yeah! I forgot to include that.

## Package coverage

In my brief exploration, RocksDB has not been updated recently in Conan ([6.20.3](https://conan.io/center/rocksdb) vs [6.27.3](https://vcpkg.info/port/rocksdb) in vcpkg), but libpq has not been updated recently in vcpkg (oh! it was just updated yesterday to [14.1](https://vcpkg.info/port/libpq) from 12.9, while Conan has [13.6](https://conan.io/center/libpq)).

vcpkg claims [1910 packages](https://vcpkg.io/en/packages.html) while Conan Center has [1238 recipes](https://github.com/conan-io/conan-center-index/tree/master/recipes). Of the packages we use, the one that vcpkg has that Conan does not is secp256k1, which we presently embed. The one that Conan has that vcpkg does not is NuDB. Both are missing ed25519-donna, which we also embed.

vcpkg has a one patch version ahead on protobuf, while Conan is one patch version ahead on soci and one minor version ahead on sqlite3.

Overall, I think they're pretty evenly matched on package availability.

## Popularity

This is hard to measure. A few data points:

- On GitHub, vcpkg (which has both the tool and the recipes) has 15.2k stars, 4.4k forks, and 1424 contributors, while conan-center-index (where people contribute recipes) has only 543 stars, 847 forks, and 515 contributors. It's easy to star and fork, but harder to merge a commit (the qualification to become a contributor).
- On merged PRs, vcpkg has 9601 while conan-center-index has 6435 and conan has 3597. So there might be fewer contributors to Conan, but they appear to be just as active.
- Conan publishes usage statistics (but [vcpkg does not](https://github.com/microsoft/vcpkg/discussions/17795)). [Here](https://blog.conan.io/2022/01/04/conan-stats-2021.html) is their report for 2021. They saw 520k downloads of the Conan client in just December, and 490k unique visitors to conan.io over the whole year. All numbers trended up in 2021.
- According to an [ISO C++ survey](https://isocpp.org/files/papers/CppDevSurvey-2021-04-summary.pdf) (slide 12), Conan and vcpkg are just about neck-and-neck in popularity, with a slight edge for Conan, but both lag far behind "I copy the code into my project", "I build them separately", and "I download prebuilt binaries". (Our community has a long way to go.)
- Both package managers have channels in the CppLang Slack. As of now, <span>#</span>conan has 2058 members while <span>#</span>vcpkg has 666. Conan also claims their channel is <span>#</span>2 by active users and <span>#</span>4 by messages.

I think both options have very strong and active communities that won't be disappearing any time soon.

**Comment by seelabs** [comment]
@thejohnfreeman Thanks for digging up the usage data. Yeah, conan still seems like a good pick.
